pkg/cmd/server/kubernetes/node.go

```diff
@@ -126,17 +126,29 @@ func (c *NodeConfig) initializeVolumeDir(ce commandExecutor, path string) (strin
 
 // RunKubelet starts the Kubelet.
 func (c *NodeConfig) RunKubelet() {
-	// TODO: clean this up and make it more formal (service named 'dns'?). Use multiple ports.
-	clusterDNS := c.KubeletConfig.ClusterDNS
-	if clusterDNS == nil {
-		if service, err := c.Client.Endpoints(kapi.NamespaceDefault).Get("kubernetes"); err == nil {
-			if ip, ok := firstIP(service, 53); ok {
-				if err := cmdutil.WaitForSuccessfulDial(false, "tcp", fmt.Sprintf("%s:%d", ip, 53), 50*time.Millisecond, 0, 2); err == nil {
-					c.KubeletConfig.ClusterDNS = net.ParseIP(ip)
+	if c.KubeletConfig.ClusterDNS == nil {
+		// Set ClusterDNS to "kubernetes" master service ip if port 53/tcp can be successfully dialed
+		if service, err := c.Client.Services(kapi.NamespaceDefault).Get("kubernetes"); err == nil {
+			if includesServicePort(service.Spec.Ports, 53, "dns-tcp") {
+				serviceIP := service.Spec.ClusterIP
+				if err := cmdutil.WaitForSuccessfulDial(false, "tcp", fmt.Sprintf("%s:%d", serviceIP, 53), 50*time.Millisecond, 0, 2); err == nil {
```

Comment:
I could see keeping the existing fallback to use port 53 on the kubernetes endpoint (regardless of name) if we can ping it. So the priority order would be:

1. configured dns ip
2. specified "dns" port 53 for service (no ping test)
3. specified "dns" port 53 for endpoint (no ping test)
4. any named port 53 for endpoint (ping test)